src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,330 @@
+            return await languageServiceBroker.RequestAsync(
+                textBuffer: textBuffer,
+                method: Methods.TextDocumentDocumentSymbolName,
+                capabilitiesFilter: (JToken x) => true,
+                languageServerName: WellKnownLspServerKinds.AlwaysActiveVSLspServer.ToUserVisibleString(),
+                parameterFactory: ParameterFactory,
+                cancellationToken: CancellationToken.None).ConfigureAwait(false);
+        }
+
+        private static string? GetFilePath(IWpfTextView textView)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+            if (textView.TextBuffer.Properties.TryGetProperty(typeof(IVsTextBuffer), out IVsTextBuffer bufferAdapter) &&
+                bufferAdapter is IPersistFileFormat persistFileFormat &&
+                ErrorHandler.Succeeded(persistFileFormat.GetCurFile(out var filePath, out _)))
+            {
+                return filePath;
+            }
+
+            return null;
+        }
+
+        private void ExpandAll(object sender, RoutedEventArgs e)
+        {
+            if (SymbolTreeInitialized)
+            {
+                var documentSymbolModels = new List<DocumentSymbolViewModel>();
+                foreach (var documentSymbolModel in SymbolsTreeItemsSource)
+                {
+                    documentSymbolModels.Add(SetIsExpanded(documentSymbolModel, true));
+                }
+
+                SymbolsTreeItemsSource = documentSymbolModels;
+                symbolTree.ItemsSource = documentSymbolModels;
+            }
+        }
+
+        private void CollapseAll(object sender, RoutedEventArgs e)
+        {
+            if (SymbolTreeInitialized)
+            {
+                var documentSymbolModels = new List<DocumentSymbolViewModel>();
+                foreach (var documentSymbolModel in SymbolsTreeItemsSource)
+                {
+                    documentSymbolModels.Add(SetIsExpanded(documentSymbolModel, false));
+                }
+
+                SymbolsTreeItemsSource = documentSymbolModels;
+                symbolTree.ItemsSource = documentSymbolModels;
+            }
+        }
+
+        private DocumentSymbolViewModel SetIsExpanded(DocumentSymbolViewModel documentSymbolModel, bool isExpanded)
+        {
+            documentSymbolModel.IsExpanded = isExpanded;
+            var documentSymbolModelChildren = new List<DocumentSymbolViewModel>();
+            foreach (var documentSymbolModelChild in documentSymbolModel.Children)
+            {
+                documentSymbolModelChildren.Add(SetIsExpanded(documentSymbolModelChild, isExpanded));
+            }
+
+            documentSymbolModel.Children = documentSymbolModelChildren;
+            return documentSymbolModel;
+        }
+
+        private void Search(object sender, EventArgs e)
+        {
+            if (SymbolsTreeItemsSource is not null)
```

Comment:
the search happens on the UI thread.  we should consider moving it off the ui thread as this could potentially takea  long time and be blocking on large files.